src/ProtocolZoo/swapping.jl

```diff
@@ -0,0 +1,188 @@
+    high_nodes = [n for n in queryall(reg, EntanglementCounterpart, pred_high, ❓; locked=false, assigned=true) if !(sync)||!isolderthan(n.slot, retention_time-buffer_time)]
+
+    (isempty(low_nodes) || isempty(high_nodes)) && return nothing
+    il = choose_low((n.tag[2] for n in low_nodes)) # TODO make [2] into a nice named property
+    ih = choose_high((n.tag[2] for n in high_nodes))
+    return (low_nodes[il], high_nodes[ih])
+end
+
+"""
+$TYPEDEF
+
+A protocol, running at a given node, that finds swappable entangled pairs and performs the swap.
+Rejects the swap candidates that are about to go past their cutoff/retention time by checking their time of creation, while the cutoff protocol, [`CutoffProt`](@ref) deletes such qubits independently.
+
+See also: [`SwapperKeeper`](@ref)
+
+
+$TYPEDFIELDS
+"""
+@kwdef struct SwapperShedder{NL,NH,CL,CH,LT} <: AbstractProtocol where {NL<:Union{Int,<:Function,Wildcard}, NH<:Union{Int,<:Function,Wildcard}, CL<:Function, CH<:Function, LT<:Union{Float64,Nothing}}
+    """time-and-schedule-tracking instance from `ConcurrentSim`"""
+    sim::Simulation
+    """a network graph of registers"""
+    net::RegisterNet
+    """the vertex of the node where swapping is happening"""
+    node::Int
+    """the vertex of one of the remote nodes for the swap, arbitrarily referred to as the "low" node (or a predicate function or a wildcard); if you are working on a repeater chain, a good choice is `<(current_node)`, i.e. any node to the "left" of the current node"""
+    nodeL::NL = ❓
+    """the vertex of the other remote node for the swap, the "high" counterpart of `nodeL`; if you are working on a repeater chain, a good choice is `>(current_node)`, i.e. any node to the "right" of the current node"""
+    nodeH::NH = ❓
+    """the `nodeL` predicate can return many positive candidates; `chooseL` picks one of them (by index into the array of filtered `nodeL` results), defaults to a random pick `arr->rand(keys(arr))`; if you are working on a repeater chain a good choice is `argmin`, i.e. the node furthest to the "left" """
+    chooseL::CL = random_index
+    """the `nodeH` counterpart for `chooseH`; if you are working on a repeater chain a good choice is `argmax`, i.e. the node furthest to the "right" """
+    chooseH::CH = random_index
+    """fixed "busy time" duration immediately before starting entanglement generation attempts"""
+    local_busy_time::Float64 = 0.0 # TODO the gates should have that busy time built in
+    """how long to wait before retrying to lock qubits if no qubits are available (`nothing` for queuing up and waiting)"""
+    retry_lock_time::LT = 0.1
+    """how many rounds of this protocol to run (`-1` for infinite))"""
+    rounds::Int = -1
+    """what is the oldest a qubit should be to be picked for a swap"""
+    retention_time::Float64 = 5.0
+    """padding time so that we don't cross the `retention_time` while the asynchronous messaging takes place"""
+    buffer_time::Float64 = 0.5
```

Comment:
It seems this protocol never uses these two separately, it only uses them together. This should be a single parameter.

Reply:
That makes sense!